### DISCUSSION TITLE
roadmap

### DISCUSSION THREAD

**Discussion by techs-sus** [headsmasher8557/tpscript#13] [Q&A]
we need a project roadmap

**Comment by techs-sus** [comment]
- better ... support
- callset supporting tuples (maybe a custom mt?)
- strings can be defined with quotes
- inner expressions such as [call $ret wait], $ret is a return value
- better virtual function documentation

**Reply by techs-sus** [reply]
we may also need to fix getindex (I don't know if it allows for arrays)

**Reply by raymonable** [reply]
new syntaxes
`double quotes ("" or " idk choose what you want) are used for telling the interpreter if it's a string`
`instructions inside [] brackets would be executed first`

for example

`setindex part Position [callset $ret Vector3.new 0 10 0]]`

**Reply by techs-sus** [reply]
reject callif;

we don't need calif because you can still emulate an if statement (although not pretty.)

the way the lua to asm compiler does this is by using a jump label + smart jmpif + set __condition $false

```
chk __bool $equ $true $false // idk if this is valid
::if_statement_0x01
log $true
set __bool $false
jmpif __bool if_statement_0x01
setstr x "done if statement"
logtxt x
```

**Comment by headsmasher8557** [comment]
i agree

**Comment by raymonable** [comment]
true

**Comment by ghost** [comment]
the issue where you can use call or loadstring to execute danger code is very bad fixing it should be added to the roadmap because my project depends on it and if someone uses the danger secure vulnerable it will combust this is so sad and it needs to be fixed its such a dangerous secure vulnerable it needs to be fixed immediately and as soon as possible because if it isn't fixed then someone might use it on my project and i'm repeating myself but you need to fix it  because its such a big issue preventing me from using tpscript and you might reply that i am contradicting myself no im not you are wrong i use it for my project and its a huge issue

**Reply by ghost** [reply]
#10

**Reply by headsmasher8557** [reply]
> be more careful with your pc then
> im not going to make a sandbox for os library

if you dont want people using `os.execute` on your project, just make it nil or something!
```
set os $nil
```

**Reply by ghost** [reply]
yeah but it can be used to execute danger code in general not just os.execute for example you could fill up the disk with garbage data and everyone will combust because the disk is full of garbage data using io or whatever you use to modify files in lua
